### DISCUSSION TITLE
Schema is auto-added when using fileuploads with no clear reason to why

### DISCUSSION THREAD

**Discussion by ghost** [fastapi/fastapi#6874] [Questions] [answered]
# Schema is auto-added when using file uploads with no clear reason to why
- View template and re-review suggestions ✅ 
- Try to find exiting issues/solutions ✅ 

## Problem:
When adding a file upload using the FastAPI's `File` and `UploadFile` two schema's seem to be automatically added without any clear reason why and with seemingly no working way to replace/overwrite/remove them. As they don't provide any meaningful information for the consumers of my API I would like to replace/remove them 😅
![image](https://user-images.githubusercontent.com/49159860/109162955-13b04780-7779-11eb-8356-d832a09f8af0.png)

## Questions
- Is this intended behaviour? (As in, shouldn't the specified responses and/or response models take precedence?)
    - No -> Ticket should be tagged with a 🐛 

- Is there a way to overwrite/remove these?
    - [Yes!](https://fastapi.tiangolo.com/advanced/extending-openapi/#modify-the-openapi-schema) (Answered by @NicolasRondon)

## Source
`src/main.py`
```py
from api.http.routers import ROUTERS
from api.settings import SETTINGS

# Setting up application logger
logging.basicConfig(level=logging.DEBUG)

# Initialize application with settings
app = FastAPI(**SETTINGS["app"])

# Add the routers
[app.include_router(router) for router in ROUTERS]
# ...
```

`src/api/http/routers/file.py`
```py
# ... imports
_dbucket = SETTINGS["backend"]["storage"]["default_bucket"]
_dfile = SETTINGS["backend"]["storage"]["default_file"]

router = APIRouter()

# ... Other routers

# router usage
@router.post(
    path="/file/upload",
    operation_id="api.http.routers.file.upload",
    tags=["Uploads & Downloads"],
    response_model=Result,
    responses=FileResponses.upload_direct)
async def upload_direct(
    name: str = _dfile, bucket: str = _dbucket, file: UploadFile = File(...)
):
    """Admin/debugging call for direct file upload"""
    if StorageClient.lookup_file(name, bucket):
        raise HTTPException(
            status_code=409, detail="The file already exists"
        )
    else:
        file.filename = name
        StorageClient.upload_direct(bucket, file)
        return Result(detail=[{"msg": "File uploaded sucessfully"}])
```

`src/api/models/generic.py`
```py
from typing import List
from pydantic import BaseModel


class Message(BaseModel):
    """Generic message model"""
    msg: str


class Result(BaseModel):
    """File/bucket generic result"""
    detail: List[Message]
```
## Additional information for reproduction
- Python version: 3.8.7 64bit
- `FileResponses` is just a container class holding responses as described [here](https://fastapi.tiangolo.com/advanced/additional-responses/) in docs
- [requirements-dump.txt](https://github.com/tiangolo/fastapi/files/6043511/requirements-dump.txt)

**Comment by NicolasRondon** [comment]
Maybe you can try [Documentation](https://fastapi.tiangolo.com/advanced/path-operation-advanced-configuration/#exclude-from-openapi)

**Comment by ghost** [comment]
I want to keep the router, but remove the auto-generated schema for the File+FileUpload* @NicolasRondon the doc reference would remove the router, but keep the auto-generated schema...?

**Comment by ghost** [comment]
When `file: UploadFile = File(...)` is called it seems to not just inject the file as expected, but to also generate and inject/append a schema for the inbound form/filelike object. That schema (as shown in the sample photo) isn't the behaviour I'd expect, nor want or need.

Kind of why my Q1 is "Is this intended behaviour?" because if it is, it should be added to the docs. Otherwise it should be filed as a bug so that it can be fixed =)

**Comment by NicolasRondon** [comment]
@alexpdr I was reading SO to answer some questions then I found this [question](https://stackoverflow.com/questions/66387641/how-to-hide-input-param-in-fastapi) You can customize the open api schema [fastapi docs](https://fastapi.tiangolo.com/advanced/extending-openapi/#modify-the-openapi-schema)

**Comment by ghost** [comment]
@NicolasRondon That worked as a temporary work-a-round 😄

I am however now convinced this is a bug and unintended behaviour from FastAPI, do you know who would be best to tag to get a `bug` label? ( @tiangolo maybe? ) That might be better than creating a new ticket as it would keep the information in one location.

```py
# Work-a-round for bad schema generation
def custom_openapi():
    """
    Overwrite auto-generated schema bug
    Ref: https://github.com/tiangolo/fastapi/issues/2870
    """
    schema = get_openapi(
        title=SETTINGS['app']['title'], 
        version=SETTINGS['app']['version'],
        routes=app.routes
    )
    del schema['components']['schemas']['Body_replace_file_file_put']
    del schema['components']['schemas']['Body_upload_direct_file_upload_post']
    return schema


# Set api generator to custom
app.openapi = custom_openapi
```

**Comment by ghost** [comment]
![image](https://user-images.githubusercontent.com/49159860/109483400-78241d00-7a7f-11eb-966d-39ec0b270802.png)
Work-a-round however breaks redoc ☹️

**Comment by ghost** [comment]
Reverted the change and debugged a littler further, the schema get's added by the `File(..)` method directly. This is seemingly as it doesn't have the Pydantic model normally used for the body (like described here: https://fastapi.tiangolo.com/tutorial/body/). 

🤔 Could there be a way to supply it with a model? (doesn't seem so via the available args, but I might be missing something)

**Comment by ghost** [comment, marked as answer]
Fix mentioned here is a working fix for the issue. 
https://stackoverflow.com/questions/60765317/how-to-create-an-openapi-schema-for-an-uploadfile-in-fastapi

Ref: #1442  

Closing issue as duplicate.

**Comment by tiangolo** [comment]
Thanks for the help here @NicolasRondon ! 👏  🙇 

Thanks for reporting back and closing the issue @alexpdr  👍

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
